build.gradle.kts

```diff
@@ -17,9 +18,25 @@ java {
     withSourcesJar()
 }
 
+modularity {
+    mixedJavaRelease(8)
+}
+
+tasks.getByName<Jar>("sourcesJar") {
+    // When the mixedJavaRelease is set, the module-info.class will be packaged into the sourcesJar
+    exclude("module-info.class")
+}
+
+tasks.compileTestJava {
+    extensions.configure<org.javamodularity.moduleplugin.extensions.CompileTestModuleOptions> {
+        isCompileOnClasspath = true
+    }
+}
```

Comment:
Why did we need to fallback to classpath mode? Does that mean that consumers of the library will have to use the classpath instead of the module path as well?

Reply:
> Why did we need to fallback to classpath mode? Does that mean that consumers of the library will have to use the classpath instead of the module path as well?

This is only for testing, and will not affect the release.

Some module related problems prevent the test, and this option can make the test work normally.

Reply:
> This is only for testing, and will not affect the release

This specific *setting* is only for testing, yes. But the fact that you had to do this makes me wary.

> Some module related problems prevent the test

Exactly. What is it? Is this problem going to affect consumers? That's exactly my question.

> this option can make the test work normally

Yes, but it seems it's just hiding the problem with the modules (unless we understand what the problem is more clearly).

Reply:
@joffrey-bion In fact, Gradle's test could not correctly reflect the module related problems.

Testing is completely different from users. The user uses the last generated jar file, and the test is based on the class files in the folder. 

Moreover, the test itself in this project has broken the module assumption —— Jigsaw does not allow different modules to contain the same package, so your tests can work normally only after they have invaded the module.

The Jigsaw problem encountered here should be caused by a bug in the gradle modules plugin. This plugin has many bugs. If the JavaFX Gradle Plugin plugin does not depends on it, I do not want to use it.

However, based on my experience working on dozens of modular Java projects, I think it is reasonable to use classpath directly. Because I haven't seen any build tool that can correctly handle Jigsaw in all cases, Gradle itself also has many module related bugs, it is difficult to avoid using workaround in projects to bypass problems related to jigsaw.

Reply:
> In fact, Gradle's test could not correctly reflect the module related problems.

It did for me in the past. Before your change, my tests had shown me I was lacking `opens` declarations for Gson for example.

> The user uses the last generated jar file, and the test is based on the class files in the folder.

By default, yes, but I think that depends on how tests are setup. I think with the Gradle Modules Plugin this default changes, because it sets up Gradle to use the module path instead. That's probably why my tests could catch problems with modules so far.

> However, based on my experience working on dozens of modular Java projects, I think it is reasonable to use classpath directly

I am not convinced that it is reasonable. A good quality library should ideally test both when the jar is used on the module path AND on the classpath.

Reply:
For some reason, despite the classpath mode it seems to pickup the problems in `module-info.java`, so it might be good enough for now 🤔